HBASE-28567.

From: Lars Francke <git@lars-francke.de>

Race condition causes MetaRegionLocationCache to never set watcher to populate meta location
---
 .../apache/hadoop/hbase/zookeeper/ZKWatcher.java   |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/hbase-zookeeper/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKWatcher.java b/hbase-zookeeper/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKWatcher.java
index 3879cb7ba9..5af7de5678 100644
--- a/hbase-zookeeper/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKWatcher.java
+++ b/hbase-zookeeper/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKWatcher.java
@@ -460,6 +460,11 @@ public class ZKWatcher implements Watcher, Abortable, Closeable {
   public List<String> getMetaReplicaNodesAndWatchChildren() throws KeeperException {
     List<String> childrenOfBaseNode =
       ZKUtil.listChildrenAndWatchForNewChildren(this, znodePaths.baseZNode);
+    // Need to throw here instead of returning an empty list if the base znode hasn't been created
+    // Caller should retry in that case, versus thinking the base znode has a watcher set
+    if (childrenOfBaseNode == null) {
+      keeperException(new KeeperException.NoNodeException(znodePaths.baseZNode));
+    }
     return filterMetaReplicaNodes(childrenOfBaseNode);
   }
 
